extensions-core/s3-extensions/src/main/java/org/apache/druid/storage/s3/output/RetryableS3OutputStream.java

```diff
@@ -103,13 +109,35 @@ public class RetryableS3OutputStream extends OutputStream
   private boolean error;
   private boolean closed;
 
+  /**
+   * A map to store number of files pending to be uploaded for a given uploadId.
+   */
+  private static final ConcurrentHashMap<String, AtomicInteger> PENDING_FILES = new ConcurrentHashMap<>();
+
+  /**
+   * A map containing the lock for a given uploadId used for notifying the main thread about the completion of
+   * s3.uploadPart() for all chunks for the uploadId, and hence starting the s3.completeMultipartUpload() for
+   * the uploadId.
+   */
+  private static final ConcurrentHashMap<String, Object> FILE_LOCK_MAP = new ConcurrentHashMap<>();
+
+  /**
+   * Semaphore to restrict the maximum number of simultaneous chunks on disk.
+   */
+  private static final int MAX_CONCURRENT_CHUNKS = 10;
```

Comment:
@cryptoe How should this be configured? Is it fine to keep it constant? Or should we add another config parameter in S3OutputConfig? Thoughts?

Reply:
This is dependent on 
* Size of the local disk we have
* SIze of the chunk
* Number of upload threads 
So we need to come up with a formula like `WorkerStorageParameters`

Reply:
@cryptoe I'm a bit confused on this. The `maxConcurrentChunks` should be the same across all instances of `RetryableS3OutputStream` (IIUC??), but `SIze of the chunk` is a per-instance property for RetryableS3OutputStream.

Similar confusion for adding a dependency on `Number of upload threads`.

Appreciate your inputs, thanks!

Reply:
@cryptoe As discussed yesterday, I've added a singleton class `S3UploadConfig` which maintains a `max(chunkSize)`, and uses that to dynamically update the `maxConcurrentNumChunks`. This is then used in `RetryableS3OutputStream` for backpressure semantics.